src/Tasks/Common/FileUtilities.MetadataReader.cs

```diff
@@ -21,31 +22,57 @@ namespace Microsoft.NET.Build.Tasks
 {
     static partial class FileUtilities
     {
+        private static Dictionary<string, (DateTime LastKnownWriteTimeUtc, Version Version)> s_versionCache = new(StringComparer.OrdinalIgnoreCase /* Not strictly correct on *nix. Fix? */);
+
         private static Version GetAssemblyVersion(string sourcePath)
         {
-            using (var assemblyStream = new FileStream(sourcePath, FileMode.Open, FileAccess.Read, FileShare.Delete | FileShare.Read))
+            // TODO: is there a way to assert that it's a full path?
+
+            DateTime lastWriteTimeUtc = File.GetLastWriteTimeUtc(sourcePath);
+
+            if (s_versionCache.TryGetValue(sourcePath, out var cacheEntry) 
+                && lastWriteTimeUtc == cacheEntry.LastKnownWriteTimeUtc)
+            {
+                return cacheEntry.Version;
+            }
+
+            Version version = GetAssemblyVersionFromFile(sourcePath);
+
+            if (version is not null)
+            {
+                s_versionCache[sourcePath] = (lastWriteTimeUtc, version);
+            }
+
+            // TODO: clear or prune cache sometimes?
```

Comment:
This is something I'd like feedback on.

I can think of a few approaches:

1. Do nothing, let it grow: the cached information is small and while it _can_ grow indefinitely in practice it won't: you're going to build the same few things over and over again within an MSBuild node lifetime.
2. Clear entirely if cache size reaches some threshold. Easy, keeps memory bounded, but perf is funky because it periodically "starts over". Also have to tune the size.
3. Keep last-used timestamp, periodically purge old entries. Complex-ish (are we starting a thread on a timer to do this?) and potentialy funky perf characteristics (if no thread, do this every Nth call?)

I think I'm talking myself into 1 here. @ladipro, thoughts?

Reply:
I agree that 1. is OK. It is a common helper so probably not safe to assume any calling pattern but the upper bound of "all assemblies on disk ever participating in any build" is reasonable given how little data is stored.